File: java/src/main/java/com/example/cote/PGMRS/신고_결과_받기.java
package com.example.cote.PGMRS;

import static java.util.stream.Collectors.*;

import java.io.*;
import java.nio.file.*;
import java.util.*;

public class 신고_결과_받기 {

  private static final int MAX_ID_LIST_LENGTH = 1000;
  private static final int MAX_REPORTS_LENGTH = 200000;
  private static final int MAX_K = 200;

  private static int[] solve(String[] idList, String[] reports, int k) {

    Map<String, Set<String>> reporter = new HashMap<>();

    for (String id : idList) {
      reporter.put(id, new HashSet<String>());
    }

    Map<String, Integer> counts = new HashMap<>();

    for (String r : reports) {
      String[] tokens = r.split(" ");
      String from = tokens[0];
      String whom = tokens[1];

      Set<String> set = reporter.get(from);
      if (set.contains(whom)) continue;

      set.add(whom);
      int c = counts.getOrDefault(whom, 0);
      counts.put(whom, c + 1);
    }

    Set<String> banned =
        counts.entrySet().stream()
            .filter(e -> e.getValue() >= k)
            .map(e -> e.getKey())
            .collect(toSet());

    return Arrays.stream(idList)
        .mapToInt(id -> (int) reporter.get(id).stream().filter(banned::contains).count())
        .toArray();
  }

  public static void main(String... args) throws IOException {

    Path p = Paths.get(System.getProperty("user.dir") + "/data/신고_결과_받기.txt");
    BufferedReader reader = Files.newBufferedReader(p);

    int testCases = Integer.parseInt(reader.readLine());

    while (testCases-- > 0) {
      String[] idList = reader.readLine().split(" ");
      int numReport = Integer.parseInt(reader.readLine());
      String[] reports = new String[numReport];

      for (int i = 0; i < numReport; i++) {
        reports[i] = reader.readLine();
      }

      int k = Integer.parseInt(reader.readLine());

      int[] answer =
          Arrays.stream(reader.readLine().split(" ")).mapToInt(Integer::parseInt).toArray();

      System.out.println("Answer : " + Arrays.toString(answer));
      System.out.println("Solved : " + Arrays.toString(solve(idList, reports, k)));
    }

    return;
  }
}

File: kotlin/src/main/kotlin/com/example/cote/PGMRS/신고_결과_받기.kt
package com.example.cote.PGMRS

import java.io.*
import java.nio.file.*

object 신고_결과_받기 {
    data class Record(val reporter: String, val accused: String)

    fun solve(
        idList: List<String>,
        records: List<List<String>>,
        k: Int
    ): List<Int> {

        val reportsById = mutableMapOf<String, MutableSet<String>>()
        val counts = mutableMapOf<String, Int>()
        val reports = records.map { Record(it[0], it[1]) }

        for ((reporter, accused) in reports) {
            val accusedByReporter =
                reportsById[reporter] ?: mutableSetOf<String>()

            if (accused in accusedByReporter) {
                continue
            }

            accusedByReporter.add(accused)
            reportsById[reporter] = accusedByReporter
            counts[accused] = (counts[accused] ?: 0) + 1
        }

        val banned =
            counts
                .asSequence()
                .filter { it -> it.value >= k }
                .map { it.key }
                .toSet()

        return idList.map { id ->
            reportsById[id]?.filter { it: String -> it in banned }?.size ?: 0
        }
    }

    fun main() {
        val reader =
            with(
                Paths.get(
                    System.getProperty("user.dir") + "/data/신고_결과_받기.txt")) {
                    Files.newBufferedReader(this)
                }

        repeat(reader.readLine().toInt()) {
            val idList = reader.readLine().split(" ")
            val records =
                IntRange(1, reader.readLine().toInt()).map {
                    reader.readLine().split(" ")
                }
            val k = reader.readLine().toInt()
            val answer = reader.readLine().split(" ").map(String::toInt)
            println("$answer ${solve(idList, records, k)}")
        }
    }
}

fun main() {
    신고_결과_받기.main()
}
